### PR TITLE
feat(ar-io-data-source): add ArIODataSource for p2p data fetching

### PR DESCRIPTION
**Pull request by karlprieb**

### CODE REVIEW

File: src/data/ar-io-data-source.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        timeout: this.requestTimeoutMs,
+      });
+
+      if (response.status !== 200) {
+        throw new Error(`Unexpected status code from peer: ${response.status}`);
+      }
+
+      return response;
+    } catch (error) {
+      log.error('Request to peer failed', { error });
```
Comment (djwhitt) [inline src/data/ar-io-data-source.ts:147]: Let's skip the try/catch here since the caller already handles the error and logs appropriately.

**Review by djwhitt — COMMENTED** [review]
Nice work! Just added a few minor suggestions.

File: tsconfig.json [hunk 2]
```diff
@@ -16,7 +16,10 @@
     "noImplicitAny": true,
     "noImplicitReturns": true,
     "noUnusedLocals": true,
-    "noUnusedParameters": true
+    "noUnusedParameters": true,
+
+    /* Temporary workaround until @ar.io/sdk is fixed */
+    "skipLibCheck": true
```
Comment (karlprieb) [inline tsconfig.json:22]: @djwhitt are you good with that?

File: src/data/ar-io-data-source.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+      // Remove node wallet from peers list if it exists
+      if (this.nodeWallet !== undefined && peers[this.nodeWallet]) {
+        delete peers[this.nodeWallet];
+      }
+
+      this.peers = peers;
+    } catch (error) {
+      log.error('Failed to update peer list', { error });
+      throw new Error('Failed to update peer list');
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/data/ar-io-data-source.ts:102]: In fact `SequentialDataSource` will catch that

File: src/data/ar-io-data-source.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    maxHopsAllowed?: number;
+    requestTimeoutMs?: number;
+    updatePeersRefreshIntervalMs?: number;
+  }) {
+    this.log = log.child({ class: 'ArIODataSource' });
+    this.nodeWallet = nodeWallet;
+    this.maxHopsAllowed = maxHopsAllowed;
+    this.requestTimeoutMs = requestTimeoutMs;
+    this.updatePeersRefreshIntervalMs = updatePeersRefreshIntervalMs;
+    this.arIO = ArIO.init();
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/data/ar-io-data-source.ts]: [`cab36d6` (#131)](https://github.com/ar-io/ar-io-node/pull/131/commits/cab36d608cc4cfd2929f6418cf34009494d7fca0)

File: src/data/ar-io-data-source.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/data/ar-io-data-source.ts]: that's great :)
[`cab36d6` (#131)](https://github.com/ar-io/ar-io-node/pull/131/commits/cab36d608cc4cfd2929f6418cf34009494d7fca0)

File: tsconfig.json (hunk 2 above)
Comment (dtfiedler) [reply tsconfig.json:22]: the issue comes from warp-contracts and `@irys/arweave` - I've pushed an issue to warp but don't expect them to handle immediately. have not diagnosed the @irys/arweave issue yet

File: src/data/ar-io-data-source.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/data/ar-io-data-source.ts:104]: Isn't this being called by `setInterval`?

File: tsconfig.json (hunk 2 above)
Comment (djwhitt) [reply tsconfig.json:22]: Yeah, this is fine for now.

File: src/data/ar-io-data-source.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (karlprieb) [reply src/data/ar-io-data-source.ts:104]: Yep, my bad. As we discussed, I'm not throwing anymore and just logging the error.

**Review by djwhitt — APPROVED** [review]

**Pull request merged**